README.md
# Flutter Autocompleter

A customizable autocomplete widget for text field for Flutter

## Features
* Supports TextField, CupertinoTextField, TextFormField and even your own custom text field widget
* Customizes list item, padding, margin and decoration
* Customizes content for loading, empty and error state
* Installs hook before and after the suggestion callback
* Sets debounce duration when listening text changes
* Auto-flip when the list is outside the viewport

## Example
```dart
Autocompleter<String>(
  debounce: 300,
  flip: true,
  controller: _textEditingController,
  callback: (q) async {
    var data = ['apple', 'banana', 'cat', 'dog', 'app', '1234'];
    return Future.value(
        data.where((element) => element.contains(q)).toList());
  },
  errorBuilder: (_) {
    return const Padding(
      padding: EdgeInsets.all(16.0),
      child: Text('Error occured'),
    );
  },
  itemBuilder: (_, s) {
    return Padding(
      padding: const EdgeInsets.all(16.0),
      child: Text(s),
    );
  },
  loadingBuilder: (_) => const Center(
        child: Padding(
          padding: EdgeInsets.all(16.0),
          child: SizedBox(
              width: 24,
              height: 24,
              child: CircularProgressIndicator(
                strokeWidth: 2,
              )),
        ),
      ),
  onTap: (s) {
    setState(() {
      _textEditingController.text = s;
      _textEditingController.selection = TextSelection.fromPosition(
          TextPosition(offset: _textEditingController.text.length));
    });
    Navigator.of(context).push(MaterialPageRoute(
        builder: (_) => Scaffold(
              body: Center(child: Text(s)),
            )));
  },
  decorationBuilder: (direction) => BoxDecoration(
          color: Colors.white,
          borderRadius: BorderRadius.vertical(
            top: direction == VerticalDirection.up
                ? const Radius.circular(16.0)
                : Radius.zero,
            bottom: direction == VerticalDirection.down
                ? const Radius.circular(16.0)
                : Radius.zero,
          ),
          boxShadow: const [
            BoxShadow(blurRadius: 3.0, color: Colors.black12)
          ]),
  child: TextField(
    controller: _textEditingController,
  ),
  direction: VerticalDirection.down)
```

CHANGELOG.md
## 0.0.3

* Update LICENSE

## 0.0.2

* Add maxHeight parameter

## 0.0.1

* Initial release


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_autocompleter_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/flutter_autocompleter.dart
import 'dart:async';
import 'dart:math';

import 'package:async/async.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_keyboard_visibility/flutter_keyboard_visibility.dart';

class Autocompleter<T> extends StatefulWidget {
  /// A customizable autocomplete widget for text field
  const Autocompleter(
      {Key? key,
      required this.controller,
      required this.callback,
      required this.itemBuilder,
      this.emptyBuilder,
      this.loadingBuilder,
      this.errorBuilder,
      this.beforeHook,
      this.afterHook,
      this.onTap,
      this.debounce = 0,
      this.maxItems,
      this.maxHeight,
      this.decorationBuilder,
      this.padding,
      this.margin,
      this.direction = VerticalDirection.down,
      this.hideOnKeyboardDismissed = true,
      this.flip = false,
      this.clipBehavior = Clip.hardEdge,
      required this.child})
      : super(key: key);

  /// [TextEditingController] of the input field.
  final TextEditingController controller;

  /// Suggestion provider that should returns a list of [T].
  final Future<List<T>> Function(String text) callback;

  /// Widget to be shown for each item in the list returned by [callback].
  final Widget Function(BuildContext context, T item) itemBuilder;

  /// Content to be shown when the list returned by [callback] is empty.
  final Widget Function(BuildContext context)? emptyBuilder;

  /// Widget to be shown when [callback] is executing.
  final Widget Function(BuildContext context)? loadingBuilder;

  /// Widget to be shown when exception is thrown while executing [callback].
  final Widget Function(BuildContext context)? errorBuilder;

  /// The hook to be invoked right before [callback] is called.
  /// Useful for performing validation, or length check before listing suggestions.
  /// Return false to stop getting suggestion and hide the suggestion list.
  final bool Function(String text)? beforeHook;

  /// The hook to be invoked right after [callback] is called.
  /// Useful for removing exact match from the list.
  /// Return false to hide the suggestion list.
  final bool Function(List<T>)? afterHook;

  /// The action when a list item is tapped.
  final void Function(T suggestion)? onTap;

  /// Delay [callback] for specific time in milliseconds. Useful for avoid frequent calls to [callback].
  final int debounce;

  /// The maximum items to be listed.
  final int? maxItems;

  /// The maximum height of the list.
  final double? maxHeight;

  /// The appearance of the suggestion list.
  final BoxDecoration Function(VerticalDirection direction)? decorationBuilder;

  /// The padding for suggestion list.
  final EdgeInsets Function(VerticalDirection direction)? padding;

  /// The margin for suggestion list.
  final EdgeInsets Function(VerticalDirection direction)? margin;

  /// The direction of the suggestion list.
  final VerticalDirection direction;

  /// Hide the suggestion list when the keyboard is dismissed.
  final bool hideOnKeyboardDismissed;

  /// Flip to the other side when no spaces left for the current [direction].
  final bool flip;

  /// The clip behavior for the suggestion list.
  final Clip clipBehavior;

  /// Widget that the suggestion list should be attached to.
  final Widget child;

  @override
  _AutocompleterState<T> createState() => _AutocompleterState<T>();
}

class _AutocompleterState<T> extends State<Autocompleter<T>>
    with WidgetsBindingObserver {
  final GlobalKey _overlayKey = GlobalKey();
  final GlobalKey _childKey = GlobalKey();
  final LayerLink _layerLink = LayerLink();
  final FocusNode _focusNode = FocusNode();
  KeyboardVisibilityController? _keyboardVisibilityController;
  StreamSubscription<bool>? _keyboardSubscription;
  ScrollController? _scrollController;

  late OverlayEntry _overlay;

  CancelableCompleter? _completer;

  /// If [_items] is null, the list would be empty without showing emptyBuilder
  List<T>? _items = [];
  bool _error = false;
  bool _loading = true;
  bool _flipped = false;
  Size? _childSize;
  Timer? _timerDebounce;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance!.addObserver(this);
    _overlay = _overlayEntry();
    widget.controller.addListener(_handleChange);
    WidgetsBinding.instance!.addPostFrameCallback((_) {
      _updateChild();
    });
    if (widget.hideOnKeyboardDismissed) {
      _keyboardVisibilityController = KeyboardVisibilityController();
      _keyboardSubscription = _keyboardVisibilityController!.onChange
          .listen(_onKeyboardVisibilityChanged);
    }
  }

  @override
  void dispose() {
    _hideAutocompleter();
    _focusNode.dispose();
    _timerDebounce?.cancel();
    WidgetsBinding.instance!.removeObserver(this);
    widget.controller.removeListener(_handleChange);
    _scrollController?.removeListener(_flipIfOverflowed);
    _keyboardSubscription?.cancel();
    super.dispose();
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (widget.flip) {
      _scrollController = Scrollable.of(context)?.widget.controller
        ?..addListener(_flipIfOverflowed);
    }
  }

  @override
  void didUpdateWidget(Autocompleter<T> oldWidget) {
    if (oldWidget.controller != widget.controller) {
      oldWidget.controller.removeListener(_handleChange);
      widget.controller.addListener(_handleChange);
    }
    if (!oldWidget.flip && widget.flip) {
      WidgetsBinding.instance?.addPostFrameCallback((_) {
        _scrollController = Scrollable.of(context)?.widget.controller
          ?..addListener(_flipIfOverflowed);
        _flipIfOverflowed();
      });
    }
    if (oldWidget.flip && !widget.flip) {
      _scrollController?.removeListener(_flipIfOverflowed);
    }
    if (oldWidget.child != widget.child) {
      WidgetsBinding.instance?.addPostFrameCallback((_) => _updateChild());
    }
    if (!oldWidget.hideOnKeyboardDismissed && widget.hideOnKeyboardDismissed) {
      _keyboardVisibilityController = KeyboardVisibilityController();
      _keyboardSubscription = _keyboardVisibilityController!.onChange
          .listen(_onKeyboardVisibilityChanged);
    }
    if (oldWidget.hideOnKeyboardDismissed && !widget.hideOnKeyboardDismissed) {
      _keyboardSubscription!.cancel();
      _keyboardSubscription = null;
      _keyboardVisibilityController = null;
    }
    if (oldWidget.beforeHook != widget.beforeHook ||
        oldWidget.afterHook != widget.afterHook) {
      WidgetsBinding.instance?.addPostFrameCallback((_) => _handleChange(true));
    }
    if (oldWidget.direction != widget.direction ||
        oldWidget.emptyBuilder != widget.emptyBuilder) {
      WidgetsBinding.instance
          ?.addPostFrameCallback((_) => _updateAutocompleter());
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  void didChangeMetrics() {
    _updateChild();
    _updateAutocompleter();
  }

  void _updateChild() {
    var child = _childKey.currentContext?.findRenderObject() as RenderBox;
    _childSize = child.size;
  }

  void _onKeyboardVisibilityChanged(bool visible) {
    if (!visible && widget.hideOnKeyboardDismissed) {
      _focusNode.unfocus();
    }
  }

  void _flipIfOverflowed() {
    bool overflowed = false;
    if (widget.flip) {
      final RenderBox? child =
          _childKey.currentContext?.findRenderObject() as RenderBox?;
      if (child != null) {
        final RenderBox? overlay =
            _overlayKey.currentContext?.findRenderObject() as RenderBox?;
        if (overlay == null) return;
        double keyboardHeight = EdgeInsets.fromWindowPadding(
                WidgetsBinding.instance!.window.viewInsets,
                WidgetsBinding.instance!.window.devicePixelRatio)
            .bottom;
        double childY = child.localToGlobal(Offset.zero).dy;
        EdgeInsets unsafeArea = EdgeInsets.fromWindowPadding(
            WidgetsBinding.instance!.window.padding,
            WidgetsBinding.instance!.window.devicePixelRatio);
        double overlayHeight = overlay.size.height;
        double screenHeight =
            WidgetsBinding.instance!.window.physicalSize.height /
                WidgetsBinding.instance!.window.devicePixelRatio;
        bool bottomOverflowed = widget.direction == VerticalDirection.down &&
            childY +
                    _childSize!.height +
                    overlayHeight +
                    keyboardHeight +
                    unsafeArea.bottom >
                screenHeight;
        bool topOverflowed = widget.direction == VerticalDirection.up &&
            childY - overlayHeight - unsafeArea.top < 0;
        if (bottomOverflowed ^ topOverflowed) {
          overflowed = true;
        } else {
          overflowed = false;
        }
      }
    }
    if (overflowed) {
      if (!_flipped) {
        _overlay.markNeedsBuild();
      }
      _flipped = true;
    } else {
      if (_flipped) {
        _overlay.markNeedsBuild();
      }
      _flipped = false;
    }
  }

  void _handleChange([bool skipDebounce = false]) async {
    _error = false;
    _loading = true;
    var newValue = widget.controller.text;
    if (widget.beforeHook != null && !widget.beforeHook!(newValue)) {
      _items = null;
      _loading = false;
      _updateAutocompleter();
      return;
    }
    if (widget.loadingBuilder != null) _updateAutocompleter();
    if (widget.debounce > 0 && !skipDebounce) {
      if (_timerDebounce?.isActive ?? false) _timerDebounce?.cancel();

      _timerDebounce = Timer(Duration(milliseconds: widget.debounce), () {
        _handleChange(true);
      });
      return;
    }

    _completer?.operation.cancel();
    _completer = CancelableCompleter<List<T>>(onCancel: () {
      _loading = false;
    });
    _loading = true;
    _completer!.complete(widget.callback(newValue));
    _completer!.operation.then((items) {
      _items = items;
      if (widget.afterHook != null && !widget.afterHook!(_items!)) {
        _items = null;
        _loading = false;
        _error = false;
        _updateAutocompleter();
        return;
      }
      _loading = false;
      _updateAutocompleter();
    }, onError: (object, stackTrace) {
      _error = true;
      _updateAutocompleter();
      if (kDebugMode) {
        print(stackTrace);
      }
    });
  }

  OverlayEntry _overlayEntry() {
    return OverlayEntry(
      builder: (BuildContext context) {
        VerticalDirection direction = (_flipped
            ? (widget.direction == VerticalDirection.down
                ? VerticalDirection.up
                : VerticalDirection.down)
            : widget.direction);
        Alignment alignment =
            Alignment(0, direction == VerticalDirection.down ? 1.0 : -1.0);
        return Positioned(
          key: _overlayKey,
          width: _childSize?.width ?? 0,
          child: CompositedTransformFollower(
            link: _layerLink,
            followerAnchor: Alignment(alignment.x, alignment.y * -1),
            targetAnchor: alignment,
            child: Material(
              color: Colors.transparent,
              child: Container(
                clipBehavior: widget.clipBehavior,
                margin:
                    widget.margin != null ? widget.margin!(direction) : null,
                constraints: widget.maxHeight != null
                    ? BoxConstraints(maxHeight: widget.maxHeight!)
                    : null,
                decoration: widget.decorationBuilder != null
                    ? widget.decorationBuilder!(direction)
                    : BoxDecoration(
                        color: Colors.white,
                        borderRadius: BorderRadius.circular(8.0),
                        boxShadow: const [
                            BoxShadow(blurRadius: 3.0, color: Colors.black12)
                          ]),
                child: _error
                    ? (widget.errorBuilder != null
                        ? widget.errorBuilder!(context)
                        : const SizedBox())
                    : (_loading && widget.loadingBuilder != null
                        ? widget.loadingBuilder!(context)
                        : (_items == null
                            ? const SizedBox()
                            : (_items!.isEmpty
                                ? (widget.emptyBuilder != null
                                    ? widget.emptyBuilder!(context)
                                    : const SizedBox())
                                : ListView.builder(
                                    shrinkWrap: true,
                                    padding: widget.padding != null
                                        ? widget.padding!(direction)
                                        : EdgeInsets.zero,
                                    itemCount: widget.maxItems != null
                                        ? min(widget.maxItems!, _items!.length)
                                        : _items!.length,
                                    itemBuilder: ((context, i) {
                                      var item = widget.itemBuilder(
                                          context, _items![i]);
                                      if (widget.onTap != null) {
                                        item = GestureDetector(
                                            behavior:
                                                HitTestBehavior.translucent,
                                            child: item,
                                            onTap: () {
                                              widget.onTap!(_items![i]);
                                              _focusNode.unfocus();
                                            });
                                      }
                                      return item;
                                    }))))),
              ),
            ),
          ),
        );
      },
    );
  }

  void _showAutocompleter() {
    if (!_overlay.mounted) Overlay.of(context)?.insert(_overlay);
  }

  void _hideAutocompleter() {
    _completer?.operation.cancel();
    if (_overlay.mounted) _overlay.remove();
  }

  void _updateAutocompleter() {
    if (!_overlay.mounted) {
      return;
    }
    _overlay.markNeedsBuild();
    WidgetsBinding.instance?.addPostFrameCallback((_) => _flipIfOverflowed());
  }

  @override
  Widget build(BuildContext context) {
    return CompositedTransformTarget(
        link: _layerLink,
        child: Container(
            key: _childKey,
            child: Focus(
              focusNode: _focusNode,
              onFocusChange: (focused) {
                if (focused) {
                  _showAutocompleter();
                } else {
                  _hideAutocompleter();
                }
              },
              child: widget.child,
            )));
  }
}


android/src/main/kotlin/com/example/flutter_autocompleter/FlutterAutocompleterPlugin.kt
package com.example.flutter_autocompleter

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterAutocompleterPlugin */
class FlutterAutocompleterPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_autocompleter")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


example/lib/demo.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter_autocompleter/flutter_autocompleter.dart';

class Demo extends StatefulWidget {
  const Demo({Key? key}) : super(key: key);

  @override
  State<Demo> createState() => _DemoState();
}

class _DemoState extends State<Demo> {
  final TextEditingController _textEditingController =
      TextEditingController(text: '');

  final TextEditingController _cupertinoTextEditingController =
      TextEditingController(text: '');

  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    _textEditingController.dispose();
    _cupertinoTextEditingController.dispose();
    super.dispose();
  }

  bool flip = false;
  VerticalDirection direction = VerticalDirection.down;
  int debounce = 0;
  Widget Function(BuildContext)? loadingBuilder;
  Widget Function(BuildContext)? emptyBuilder;
  bool Function(String)? beforeHook;

  Widget _autocompleter(BuildContext context, Widget child,
      TextEditingController textEditingController) {
    return Autocompleter<String>(
        debounce: debounce,
        flip: flip,
        controller: textEditingController,
        callback: (q) async {
          var data = ['apple', 'banana', 'cat', 'dog', 'app', '1234'];
          await Future.delayed(const Duration(seconds: 1));
          return Future.value(
              data.where((element) => element.contains(q)).toList());
        },
        errorBuilder: (_) {
          return const Padding(
            padding: EdgeInsets.all(16.0),
            child: Text('Error occured'),
          );
        },
        itemBuilder: (_, s) {
          return Padding(
            padding: const EdgeInsets.all(16.0),
            child: Text(s),
          );
        },
        loadingBuilder: loadingBuilder,
        emptyBuilder: emptyBuilder,
        onTap: (s) {
          setState(() {
            textEditingController.text = s;
            textEditingController.selection = TextSelection.fromPosition(
                TextPosition(offset: textEditingController.text.length));
          });
          Navigator.of(context).push(MaterialPageRoute(
              builder: (_) => Scaffold(
                    body: Center(child: Text(s)),
                  )));
        },
        beforeHook: beforeHook,
        decorationBuilder: (direction) => BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.vertical(
                  top: direction == VerticalDirection.up
                      ? const Radius.circular(16.0)
                      : Radius.zero,
                  bottom: direction == VerticalDirection.down
                      ? const Radius.circular(16.0)
                      : Radius.zero,
                ),
                boxShadow: const [
                  BoxShadow(blurRadius: 3.0, color: Colors.black12)
                ]),
        child: child,
        direction: direction);
  }

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.start,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: <Widget>[
          options(context),
          Padding(
              padding: const EdgeInsets.all(16),
              child: Column(children: [
                _autocompleter(
                    context,
                    TextField(
                      controller: _textEditingController,
                    ),
                    _textEditingController),
                const SizedBox(
                  height: 16,
                ),
                _autocompleter(
                    context,
                    CupertinoTextField(
                      controller: _cupertinoTextEditingController,
                    ),
                    _cupertinoTextEditingController)
              ])),
        ],
      ),
    );
  }

  List<bool Function(String)?> beforeHookOptions = [
    null,
    (s) => s.length > 1,
    (s) {
      return num.tryParse(s) != null;
    }
  ];
  Widget options(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(16.0),
      child: Wrap(
        spacing: 16,
        runSpacing: 16,
        children: [
          Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              const Text('Direction'),
              const SizedBox(
                width: 16,
              ),
              ToggleButtons(
                children: const [
                  Text('Down'),
                  Text('Up'),
                ],
                onPressed: (int index) {
                  setState(() {
                    direction = index == 0
                        ? VerticalDirection.down
                        : VerticalDirection.up;
                  });
                },
                isSelected: [
                  direction == VerticalDirection.down,
                  direction == VerticalDirection.up
                ],
              ),
            ],
          ),
          Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              const Text('Flip'),
              const SizedBox(
                width: 16,
              ),
              ToggleButtons(
                children: const [
                  Text('True'),
                  Text('False'),
                ],
                onPressed: (int index) {
                  setState(() {
                    flip = index == 0;
                  });
                },
                isSelected: [flip, !flip],
              ),
            ],
          ),
          Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              const Text('Debounce'),
              const SizedBox(
                width: 16,
              ),
              ToggleButtons(
                children: const [
                  Text('0'),
                  Text('1000'),
                  Text('2000'),
                ],
                onPressed: (int index) {
                  setState(() {
                    debounce = [0, 1000, 2000][index];
                  });
                },
                isSelected: [0 == debounce, 1000 == debounce, 2000 == debounce],
              ),
            ],
          ),
          Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              const Text('Loading builder'),
              const SizedBox(
                width: 16,
              ),
              ToggleButtons(
                children: const [
                  Text('No'),
                  Text('Circular Progress Indicator'),
                ],
                onPressed: (int index) {
                  setState(() {
                    loadingBuilder = [
                      null,
                      (context) => const Center(
                            child: Padding(
                              padding: EdgeInsets.all(16.0),
                              child: SizedBox(
                                  width: 24,
                                  height: 24,
                                  child: CircularProgressIndicator(
                                    strokeWidth: 2,
                                  )),
                            ),
                          ),
                    ][index];
                  });
                },
                isSelected: [loadingBuilder == null, loadingBuilder != null],
              ),
            ],
          ),
          Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              const Text('Empty builder'),
              const SizedBox(
                width: 16,
              ),
              ToggleButtons(
                children: const [
                  Text('No'),
                  Text('Text'),
                ],
                onPressed: (int index) {
                  setState(() {
                    emptyBuilder = [
                      null,
                      (context) => const Center(
                          child: Padding(
                              padding: EdgeInsets.all(16.0),
                              child: Text('No results'))),
                    ][index];
                  });
                },
                isSelected: [emptyBuilder == null, emptyBuilder != null],
              ),
            ],
          ),
          Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              const Text('Before hook'),
              const SizedBox(
                width: 16,
              ),
              ToggleButtons(
                children: const [
                  Text('No'),
                  Text('Length > 1'),
                  Text('Number only'),
                ],
                onPressed: (int index) {
                  setState(() {
                    beforeHook = beforeHookOptions[index];
                  });
                },
                isSelected: List.generate(
                    3, (index) => beforeHookOptions[index] == beforeHook),
              ),
            ],
          ),
        ],
      ),
    );
  }
}
